server/src/main/java/com/objectcomputing/checkins/services/action_item/ActionItemServicesImpl.java

```diff
@@ -1,79 +1,121 @@
         }
+
         return actionItemRet;
+
     }
 
     public ActionItem read(@NotNull UUID id) {
-        return actionItemRepo.findById(id).orElse(null);
 
-    }
+        String workEmail = securityService != null ? securityService.getAuthentication().get().getAttributes().get("email").toString() : null;
+        MemberProfile currentUser = workEmail != null ? currentUserServices.findOrSaveUser(null, workEmail) : null;
+        Boolean isAdmin = securityService != null && securityService.hasRole(RoleType.Constants.ADMIN_ROLE);
+
+        ActionItem actionItemResult = actionItemRepo.findById(id).orElse(null);
+
+        validate(actionItemResult == null, "Invalid action item id %s", id);
+
+        if (!isAdmin) {
+            CheckIn checkinRecord = checkinRepo.findById(actionItemResult.getCheckinid()).orElse(null);
+            final UUID pdlId = checkinRecord != null ? checkinRecord.getPdlId() : null;
+            final UUID createById = checkinRecord != null ? checkinRecord.getTeamMemberId() : null;
+            validate(!currentUser.getId().equals(pdlId) && !currentUser.getId().equals(createById), "User is unauthorized to do this operation");
+        }
+
+        return actionItemResult;
 
-    public Set<ActionItem> readAll() {
-        Set<ActionItem> actionItems = new HashSet<>();
-        actionItemRepo.findAll().forEach(actionItems::add);
-        return actionItems;
     }
 
     public ActionItem update(ActionItem actionItem) {
         ActionItem actionItemRet = null;
+        String workEmail = securityService != null ? securityService.getAuthentication().get().getAttributes().get("email").toString() : null;
+        MemberProfile currentUser = workEmail != null ? currentUserServices.findOrSaveUser(null, workEmail) : null;
+        Boolean isAdmin = securityService != null && securityService.hasRole(RoleType.Constants.ADMIN_ROLE);
+
         if (actionItem != null) {
             final UUID id = actionItem.getId();
-            final UUID guildId = actionItem.getCheckinid();
+            final UUID checkinId = actionItem.getCheckinid();
             final UUID createById = actionItem.getCreatedbyid();
-            if (guildId == null || createById == null) {
-                throw new ActionItemBadArgException(String.format("Invalid actionItem %s", actionItem));
-            } else if (id == null || !actionItemRepo.findById(id).isPresent()) {
-                throw new ActionItemBadArgException(String.format("Unable to locate actionItem to update with id %s", id));
-            } else if (!checkinRepo.findById(guildId).isPresent()) {
-                throw new ActionItemBadArgException(String.format("CheckIn %s doesn't exist", guildId));
-            } else if (!memberRepo.findById(createById).isPresent()) {
-                throw new ActionItemBadArgException(String.format("Member %s doesn't exist", createById));
+
+            CheckIn checkinRecord = checkinRepo.findById(checkinId).orElse(null);
```

Comment:
* Asking for a friend

Not to throw Holly under the bus, but what's the best practice when we call one service from another? Is is okay to call database layer of another service directly? or is it good practice to call service (interface)? or it doesn't matter?

@meyerinj @mkimberlin @bhatiam12

Reply:
The purpose of the service layer is to abstract some back end process (in this case the repositories) from some calling process. There's no universal answer on whether separate services should interact with separate repositories, but the structure of check-ins being as highly delineated as it is I would generally lean toward going through the services rather than directly to the repositories. This will serve to prevent duplication of input validation and force exposure of any issues of permissions being changed in one area but not another.